concepts/functions/introduction.md

```diff
@@ -0,0 +1,119 @@
+# Introduction
+
+## Terminology
+
+You have already used and written C++ functions in a couple of concepts.
+It is time to get technical.
+The code snippet below shows the most common terms for easy reference.
+As C++ ignores whitespace, the formatting has been changed to put each element on a single line.
+
+```cpp
+// Function declaration:
+bool admin_detected(string user, string password) // Function header
+;                                                 // Don't forget the ';' for the declaration
+
+// Function definition:
+bool                                              // Return type
+admin_detected                                    // Function name
+(string user, string password)                    // Parameter list
+{ return user == "admin" && password == "1234"; } // Function body
+```
+~~~~exercism/advanced
+The declaration works like a note to the compiler, that there is a function of that name, return type and parameter list.
+The code will not work, if the definition is missing.
+Declarations are optional, they are needed if you use the function before its definiton.
+Declarations can solve problems like cyclic references and they can be used to separate the interface from the implementation.
+~~~~
+
+## The const qualifier
+
+Sometimes you want to make sure that values cannot be changed after they have been initialized.
+C++ uses the `const` keyword as a qualifier for constants.
+
+```cpp
+const int number_of_dragon_balls{7};
+number_of_dragon_balls--; // compilation error
+```
+
+~~~~exercism/note
+You will often see constants written in _UPPER_SNAKE_CASE_.
+It is recommened to reserve this casing for macros, if there is no other convention.
+~~~~
+
+If you try to change a constant variable after it has been set, your code will not compile.
+This helps to avoid unintended changes but also opens up optimization possibilities for the compiler.
+As a human, it is also easier to reason about the code, if you know that certain parts will not be affected.
+
+You can also have `const` as a qualifier for function parameters.
+
+```cpp
+string guess_number(const int& secret, const int& guess) {
+    if (secret < guess) return "lower.";
+    if (secret > guess) return "higher.";
+    return "exact!";
+}
+```
+
+When you pass a `const` reference to the function, you can be sure that it will be left unchanged.
+You will often see `const` references for objects that might be costly to copy, like longer strings.
+A third use-case for the `const` qualifier are member functions that do not change the instance of a class.
+
+```cpp
+class Stubborn {
+    public:
+    Stubborn(string reply) {
+        response = reply;
+    }
+    string answer() const {
+        return response;
+    }
+    private:
+    string response{};
+};
+```
```

Comment:
another thing I missed: Before the snippet you explain that `const` references are often used for objects that are expensive to copy. You could consider showing that in the snippet.